examples/structured_2d_dgsem/elixir_shallowwater_well_balanced_wet_dry.jl

```diff
@@ -0,0 +1,235 @@
+# Callbacks
+
+summary_callback = SummaryCallback()
+
+analysis_interval = 1000
+analysis_callback = AnalysisCallback(semi, interval=analysis_interval, save_analysis=false)
+
+alive_callback = AliveCallback(analysis_interval=analysis_interval)
+
+save_solution = SaveSolutionCallback(interval=1000,
+                                     save_initial_solution=true,
+                                     save_final_solution=true)
+
+stepsize_callback = StepsizeCallback(cfl=1.0)
+
+callbacks = CallbackSet(summary_callback, analysis_callback, alive_callback, save_solution, stepsize_callback)
+
+stage_limiter! = PositivityPreservingLimiterShallowWater(thresholds=(equations.threshold_limiter,),
+                                                         variables=(Trixi.waterheight,))
+
+###############################################################################
+# run the simulation
+
+sol = solve(ode, SSPRK43(stage_limiter!), dt=1.0,
+            save_everystep=false, callback=callbacks, adaptive=false);
+
+summary_callback() # print the timer summary
+
+###############################################################################
+# Workaround to compute the well-balancedness error for this partiular problem
+# that has two reference water heights. One for a lake to the left of the
+# discontinuous bottom topography `H0_upper = 2.5` and another for a lake to the
+# right of the discontinuous bottom topography `H0_lower = 1.5`.
+
+# Declare a special version of the function to compute the lake-at-rest error
+# OBS! The reference water height values are hardcoded for convenience.
+function lake_at_rest_error_two_level(u, element_id, equations::ShallowWaterEquations2D)
+  h, _, _, b = u
+
+  # For well-balancedness testing with possible wet/dry regions the reference
+  # water height `H0` accounts for the possibility that the bottom topography
+  # can emerge out of the water as well as for the threshold offset to avoid
+  # division by a "hard" zero water heights as well.
+
+  # very brute force approach to recovery the appropriate reference water height
+  # on the StructuredMesh
+  if ((element_id >= 1   && element_id <= 8  ) ||
+      (element_id >= 17  && element_id <= 24 ) ||
+      (element_id >= 33  && element_id <= 40 ) ||
+      (element_id >= 49  && element_id <= 56 ) ||
+      (element_id >= 65  && element_id <= 72 ) ||
+      (element_id >= 81  && element_id <= 88 ) ||
+      (element_id >= 97  && element_id <= 104) ||
+      (element_id >= 113 && element_id <= 120) ||
+      (element_id >= 129 && element_id <= 136) ||
+      (element_id >= 145 && element_id <= 152) ||
+      (element_id >= 161 && element_id <= 168) ||
+      (element_id >= 177 && element_id <= 184) ||
+      (element_id >= 193 && element_id <= 200) ||
+      (element_id >= 209 && element_id <= 216) ||
+      (element_id >= 225 && element_id <= 232) ||
+      (element_id >= 241 && element_id <= 248))
```

Comment:
Is it really necessary to use the element IDs instead of the spatial coordinate? If we could use the latter, we could use the standard integration functionality...
If there is really no other way, this is okay to keep in the elixir, I think

Reply:
The problem is always that the LGL nodes include the endpoints. So if we have something like
```julia
if x[1] <= 0.5
   h = 10
else # x[1] > 0.5
   h = 16
end
```
then all the LGL nodes in the left element will be initialized with h=10 but the initialization in the right element would be h=10 for the first LGL node and h=16 for all the other LGL nodes. So even when a discontinuity is aligned with an element interface it is difficult to set a "true" discontinuity with just a single if statement.

Reply:
As far as I understood, we need the element IDs. If we did this using the spatial coordinate, the interface values in both elements would be set to the same value as their spatial coordinate in the domain was the same. Then, there would be no real discontinuity. But, the handling of discontinuities is the interesting part that should be tested. Especially in the ParaView output, that does not look good.
I thought about some other structure using a for loop to mimic the Boolean evaluation. At leas, it would save several lines of code, especially because it can be used twice in the elixir. If wanted, I can try to implement this

Reply:
I need to to keep experimenting. On straight-sided elements this might be possible but I do not hold out much hope for curvilinear elements.

Reply:
Would this issue be another possible application/motivation for having the ability to handle variable coefficients?

Reply:
Update: Yes, it does work to use `prevfloat` and `nextfloat` to set a true discontinuity at a straight-sided discontinuous interface. I still kept this process separate with special `initial_condition` and error analysis functions for two reasons:

1. This setup of a discontinuity in the x-direction is very specific to this test case. The strategy of applying `prevfloat` and `nextfloat` will probably work in general for straight-sided discontinuity initialization. However, it seems like overkill to change the call to initial conditions in the core part of Trixi to account for such a special setup.
2. The computation of the `lake_at_rest_error` for this test case is still special because there are two reference heights. So there should still be the special workaround at the end of the file to compute this error. It seems like a strange idea to design functionality and call signatures for a single test that is only used for debugging purposes and making sure that the methods work for a complicated test case.

Reply:
> Would this issue be another possible application/motivation for having the ability to handle variable coefficients?

I think this is an orthogonal issue. Here the problem is really: given a single point `x` can you decide / determine how to initialize a discontinuity. This is more related to the LGL nodes and the "double" point at a discontinuous interface across straight-sided elements. For curvilinear / unstructured meshes setting a "true" discontinuity is even more complicated and this `prevfloat` and `nextfloat` strategy will not work unless you know the curvature of the boundary with which to compare against.

Reply:
Maybe we should discuss this again at the next Trixi.jl meeting?

Reply:
Sure, we can add it to the agenda.

Reply:
This strategy of setting true discontinuities will be a subject to a separate PR as discussed at the Trixi.jl meeting.